### ISSUE TITLE
Codition of ' WhenPresent' has NPE problem

### ISSUE THREAD

**Issue by 411937647** [mybatis/mybatis-dynamic-sql#171]
Like 'IsNotInWhenPresent' or ‘IsInWhenPresent’  witch param is type collection.
When pass a value null instead of empty collection will cause NPE.
`java.lang.NullPointerException
		at java.util.Objects.requireNonNull(Objects.java:203)
		at org.mybatis.dynamic.sql.AbstractListValueCondition.<init>(AbstractListValueCondition.java:34)
		at org.mybatis.dynamic.sql.where.condition.IsNotIn.<init>(IsNotIn.java:30)`
Because of there codition's construction method use **collection.stream.filter()** to judge whether or not to render,when collection is null  NPE problem show.
`  protected IsInWhenPresent(Collection<T> values) {
        super(values, s -> s.filter(Objects::nonNull));
    }`
I had to initialization value with empty list like this.
`   private List<Integer> departIdList = Collections.emptyList();

    private List<String> excludeAdminRoleList = Collections.emptyList();`

Would you consider change judging condition like
 '  protected IsInWhenPresent(Collection<T> values) {
        super(values, s ->null != s && s.stream.filter(Objects::nonNull));
    } ' or find a best way to solve this situation ?

**Comment by jeffgbutler** [comment]
I have a strong opinion is that it is a good idea to avoid `if` statements when possible. Making lists non-null is part of that. If we were to allow a null list, then it forces two `if` statements:

1. The library's code needs to check for null
1. Your code needs to change to make a decision about whether to create the list or not

Why not always create the list and add elements when needed? This lowers the complexity for everyone.

BTW - I recognize that `Objects.requireNonNull` is an embedded `if` statement. My only defense is that it is consistently used throughout the library. Using it consistently in constructors allows the library to avoid null issues in the remainder of the code. If Java ever implements a consistent null handling paradigm like Kotlin then it will be an easy change to adopt it.

**Comment by jeffgbutler** [comment]
I've been thinking that the name of these conditions is misleading and might be causing confusion. The "in" conditions work differently than the other conditions in that the "in" clause is only rendered if the list contains values. Conditions like "IsInWhenPresent" apply a non-null filter to the **values** in the list, not to the list itself.

So perhaps a condition name like "IsInIgnoringNulls" might make more sense. What do you think?

**Comment by 411937647** [comment]
The naming thing really bothered me for a while, and it wasn't until a BUG occurred that I realized it wouldn't be null.
I think it's better to start by following the code specification and try to avoid null elements.
'IsInIgnoringNulls' is a clear and intuitive name, and I've used' IsInWhenPresent 'or' IsNotInWhenPresent 'in many parts of my project, and changing the name now has a big impact on the project.
Thanks for your reply. My problem have been solved
